File: Africa 2010, Qualification Round/ProblemCT9Spelling.py
'''
Created on 02/08/2014

@author: miguel
'''

if __name__ == '__main__':
    
    myKeyboard = {'a': 2, 'b': 22, 'c': 222, 'd': 3, 'e': 33, 'f': 333, 'g': 4, 'h': 44, 'i': 444, 'j': 5, 'k': 55, 'l': 555, 'm': 6, 'n': 66, 'o': 666, 'p': 7, 'q': 77, 'r': 777, 's': 7777,'t': 8, 'u': 88, 'v': 888, 'w': 9, 'x':99, 'y': 999, 'z': 9999}

    numberOfCases = int(input())
    f = open("output.txt", 'w')
    for i in range(numberOfCases):
        answer = "Case #"+str(i+1)+": "
        message = list(input())
        for j in message:
            if not j in myKeyboard:
                if(answer.endswith('0')):
                    answer += ' 0'
                else:
                    answer += '0'
            else:
                if(answer.endswith(str(myKeyboard[j])[:1])): # Se answer acaba com a primeira letra de myKeyboard[j]
                    answer += ' ' + str(myKeyboard[j])
                else:
                    answer += str(myKeyboard[j])
        f.write(answer+'\n')

File: Africa 2010, Qualification Round/ProblemCT9Spelling.java
import java.io.FileNotFoundException;
import java.io.PrintWriter;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;


public class ProblemCT9Spelling {

	public static void main(String[] args) throws FileNotFoundException {
		Scanner sc = new Scanner(System.in);
		PrintWriter file = new PrintWriter("output.txt");
		Map<Character, Integer> keyboard = new HashMap<>();
		keyboard.put('a', 2);
		keyboard.put('b', 22);
		keyboard.put('c', 222);
		keyboard.put('d', 3);
		keyboard.put('e', 33);
		keyboard.put('f', 333);
		keyboard.put('g', 4);
		keyboard.put('h', 44);
		keyboard.put('i', 444);
		keyboard.put('j', 5);
		keyboard.put('k', 55);
		keyboard.put('l', 555);
		keyboard.put('m', 6);
		keyboard.put('n', 66);
		keyboard.put('o', 666);
		keyboard.put('p', 7);
		keyboard.put('q', 77);
		keyboard.put('r', 777);
		keyboard.put('s', 7777);
		keyboard.put('t', 8);
		keyboard.put('u', 88);
		keyboard.put('v', 888);
		keyboard.put('w', 9);
		keyboard.put('x', 99);
		keyboard.put('y', 999);
		keyboard.put('z', 9999);
		
		int numberOfCases = 0;
		int limit= sc.nextInt();
		sc.nextLine();
		while(++numberOfCases < limit+1){
			String answer = "Case #"+numberOfCases+": ";
			char[] message = sc.nextLine().toCharArray(); //Turns the string into a char array
			for(int character = 0; character < message.length; character++){
				if(keyboard.get(message[character]) != null){ //If the character is not a space
					char lastChar = keyboard.get(message[character]).toString().charAt(0);
					if(answer.endsWith(""+lastChar)){ //If the character is in the same key as the previous letter
						answer += " " + keyboard.get(message[character]); //Add a space
					}
					else { //If it is not the same
						answer += keyboard.get(message[character]);
					}
				} 
				else if(message[character] == ' '){ //If the character is a space
					if(answer.endsWith("0")){
						answer += " 0";
					}
					else { // If it is not a space
						answer += '0';
					}
				}
			}
			file.println(answer);
		}
		file.close();
		sc.close();
	}
}
